app/pages/challenges/components/challengeForm.tsx

```diff
@@ -20,24 +21,103 @@ const challengeSchema = z.object({
   participants: z.array(z.number()).optional(),
 });
 
+interface FriendOption {
+  id: number;
+  name: string;
+}
+
 interface ChallengeFormProps {
   onSubmit: (data: ChallengeRequest) => void;
   addError?: Error | null;
+  userId: number;
 }
 
 export const ChallengeForm: React.FC<ChallengeFormProps> = ({
   onSubmit,
   addError,
+  userId,
 }) => {
   const [formData, setFormData] = useState({
     title: "",
     description: "",
     goal: "",
     unit: "",
   });
-  const [friendIds, setFriendIds] = useState<string>("");
-  const [invitedFriends, setInvitedFriends] = useState<string[]>([]);
+  const [invitedFriends, setInvitedFriends] = useState<FriendOption[]>([]);
   const [errors, setErrors] = useState<Record<string, string>>({});
+  const [isSearchOpen, setIsSearchOpen] = useState(false);
+  const [searchTerm, setSearchTerm] = useState("");
+  const searchRef = useRef<HTMLDivElement>(null);
+
+  const [friends, setFriends] = useState<FriendOption[]>([]);
+  const [isLoading, setIsLoading] = useState(false);
+  const [loadError, setLoadError] = useState<string | null>(null);
+
+  useEffect(() => {
+    const fetchFriends = async () => {
+      setIsLoading(true);
+      setLoadError(null);
+
+      try {
+        const response = await apis().friends.getFriends(userId);
+        if (response.status !== 200) {
+          throw new Error(`Error fetching friends: ${response.statusText}`);
+        }
+
+        const friendIds: number[] = await response.data;
+        const friendProfiles = await Promise.all(
+          friendIds.map(async (id) => {
+            try {
+              const profileResponse = await apis().user.getUserProfile(id);
+              return {
+                id: profileResponse.data.id,
+                name: profileResponse.data.name,
+                imageUrl: profileResponse.data.imageUrl,
+              };
+            } catch (error) {
+              console.error(`Failed to fetch profile for user ${id}:`, error);
+              return null;
+            }
+          }),
+        );
+
+        setFriends(friendProfiles.filter(Boolean) as FriendOption[]);
+      } catch (error) {
+        console.error("Failed to fetch friends:", error);
+        setLoadError(
+          error instanceof Error ? error.message : "Failed to load friends",
+        );
+      } finally {
+        setIsLoading(false);
+      }
+    };
+
+    if (userId) {
+      fetchFriends();
+    }
+  }, [userId]);
+
+  const filteredFriends = friends.filter(
+    (friend) =>
+      friend.name.toLowerCase().includes(searchTerm.toLowerCase()) &&
```

Comment:
use debouncing https://kettanaito.com/blog/debounce-vs-throttle

this is a standard practice, instead of sending request every key press of a user just wait 500ms until he ends typing and only then send a request